pkg/pipeline-selector/pipeline-selector.go

```diff
@@ -0,0 +1,154 @@
+	for _, pipelineSelector := range selectors.Spec.Selectors {
+		if pipelineConditionsMatchComponentParameters(&pipelineSelector.WhenConditions, selectionParameters) {
+			return &pipelineSelector.PipelineRef, pipelineSelector.PipelineParams
+		}
+	}
+	return nil, nil
+}
+
+// pipelineConditionsMatchComponentParameters evaluates given pipeline selector against component parameters.
+// In other words, checks if given pipeline can build the component (according to what the pipeline conditions say).
+func pipelineConditionsMatchComponentParameters(pipeline, component *buildappstudiov1alpha1.WhenCondition) bool {
+	if pipeline.Language != "" && !pipelineMatchesComponentCondition(pipeline.Language, component.Language) {
+		return false
+	}
+	if pipeline.ProjectType != "" && !pipelineMatchesComponentCondition(pipeline.ProjectType, component.ProjectType) {
+		return false
+	}
+
+	if pipeline.DockerfileRequired != "" {
+		if buildappstudiov1alpha1.IsBoolStringTrue(pipeline.DockerfileRequired) != buildappstudiov1alpha1.IsBoolStringTrue(component.DockerfileRequired) {
+			return false
+		}
+	}
+
+	if pipeline.ComponentName != "" && !pipelineMatchesComponentCondition(pipeline.ComponentName, component.ComponentName) {
+		return false
+	}
+
+	if len(pipeline.Labels) != 0 && !pipelineMatchesComponentLabels(pipeline.Labels, component.Labels) {
+		return false
+	}
+	if len(pipeline.Annotations) != 0 && !pipelineMatchesComponentLabels(pipeline.Annotations, component.Annotations) {
+		return false
+	}
+
+	return true
+}
+
+// pipelineMatchesComponentCondition checks if component condition is covered by the pipeline conditions.
+// For example, component condition (for the language key) is "java", pipeline conditipons are "python,java,nodejs", result is true.
+func pipelineMatchesComponentCondition(pipelineConditions, componentCondition string) bool {
+	componentCondition = strings.ToLower(strings.TrimSpace(componentCondition))
+
+	values := strings.Split(pipelineConditions, ",")
+	for _, value := range values {
+		value = strings.ToLower(strings.TrimSpace(value))
+		if componentCondition == value {
+			return true
+		}
+	}
+
+	return false
+}
+
+// pipelineMatchesComponentLabels checks if given pipeline supports build of the the component by looking at labels.
+// For example, component labels are:
+//
+//	appstudio/builder: maven
+//	anotherLable: someValue
+//
+// supporetd by the pipeline labels are:
+//
+//	appstudio/builder: maven,gradle
+//
+// The result is true.
+func pipelineMatchesComponentLabels(pipelineLabels, componentLabels map[string]string) bool {
+	for labelName, labelSupportedValues := range pipelineLabels {
+		if componentLabelValue, componentLabelExists := componentLabels[labelName]; componentLabelExists {
+			if !pipelineMatchesComponentCondition(labelSupportedValues, componentLabelValue) {
+				return false
+			}
+		} else {
+			return false
```

Comment:
This does not look right? From your example above if `anotherLable: someValue` was the first item in the loop then `componentLabelExists` would return false, even though you are expecting a true result.

Reply:
`anotherLable: someValue` will never be in the top loop because it's label on the component, but we iterate over labels on the pipeline condition. In other words, the component must match all the labels on the pipeline condition, but it can have others, unrelated to the pipeline, labels as well.

Reply:
Ah, I gave a bad example, but it you have any unrelated label on the pipeline then it will fail to match, which I am not sure is what you want.

E.g. if the pipeline had `unrelated: someValue` then this will cause the match to fail, which basically means that adding any unrelated labels/annotations to a pipeline can change the builder that is selected. Is this really the logic that you are after as this does not really feel right?

Reply:
Ah, thanks for the example.